File: go/pkg/lib/mbt_communication.go
package lib

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
	"go.uber.org/zap"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/emptypb"
)

// MbtMessage defines the structure for the message body.
type MbtMessage struct {
	Type string `json:"type"`
	Body string `json:"body"`
}

// Queue this function is dedicated to
const (
	mbtTestingQueueName = "mbt_testing_queue"
	rabbitPort          = "5672"
	rabbitDNS           = "rabbitmq.core.svc.cluster.local"
)

func SendToTestQueue(
	ctx context.Context,
	msgType string,
	msgBody interface{}, // expected to be a proto.Message
) *emptypb.Empty {
	logger.Debug("Starting SendToTestQueue", zap.String("queue", mbtTestingQueueName))

	connectionString, err := getAMQConnectionString()
	if err != nil {
		logger.Error("Failed to get RabbitMQ connection string", zap.Error(err))
		return &emptypb.Empty{}
	}

	conn, err := amqp.Dial(connectionString)
	if err != nil {
		logger.Error("Failed to connect to RabbitMQ", zap.Error(err))
		return &emptypb.Empty{}
	}
	defer conn.Close()

	channel, err := conn.Channel()
	if err != nil {
		logger.Error("Failed to open channel", zap.Error(err))
		return &emptypb.Empty{}
	}
	defer channel.Close()

	if !queueExists(channel, mbtTestingQueueName) {
		logger.Error("Queue does not exist", zap.String("queue", mbtTestingQueueName))
		return &emptypb.Empty{}
	}

	pbMsg, ok := msgBody.(proto.Message)
	if !ok {
		logger.Error("msgBody is not a proto.Message")
		return &emptypb.Empty{}
	}

	binaryBody, err := proto.Marshal(pbMsg)
	if err != nil {
		logger.Error("Failed to marshal protobuf message", zap.Error(err))
		return &emptypb.Empty{}
	}

	payload := MbtMessage{
		Type: msgType,
		Body: base64.StdEncoding.EncodeToString(binaryBody),
	}

	bodyBytes, err := json.Marshal(payload)
	if err != nil {
		logger.Error("Failed to marshal payload to JSON", zap.Error(err))
		return &emptypb.Empty{}
	}

	publishing := amqp.Publishing{
		ContentType: "application/json",
		Body:        bodyBytes,
	}

	ctxWithTimeout, cancel := context.WithTimeout(ctx, 8*time.Second)
	defer cancel()

	logger.Info("Publishing message",
		zap.String("queue", mbtTestingQueueName),
		zap.String("type", msgType),
	)

	err = channel.PublishWithContext(
		ctxWithTimeout,
		"",                  // exchange
		mbtTestingQueueName, // routing key
		true,                // mandatory
		false,               // immediate
		publishing,
	)
	if err != nil {
		logger.Error("Failed to publish message",
			zap.String("queue", mbtTestingQueueName),
			zap.Error(err),
		)
	}

	logger.Debug("Finished SendToTestQueue", zap.String("queue", mbtTestingQueueName))
	return &emptypb.Empty{}
}

func getAMQConnectionString() (string, error) {
	user := os.Getenv("AMQ_USER")
	pw := os.Getenv("AMQ_PASSWORD")

	return fmt.Sprintf("amqp://%s:%s@%s:%s/", user, pw, rabbitDNS, rabbitPort), nil
}

func queueExists(ch *amqp.Channel, queueName string) bool {
	_, err := ch.QueueDeclarePassive(
		queueName, // name
		true,      // durable
		false,     // delete when unused
		false,     // exclusive
		false,     // no-wait
		nil,       // args
	)
	return err == nil
}

File: python/dynamos-python-lib/dynamos/mbt_communication.py
import base64
import json
import threading
import pika
import os

class TestQueuePublisher:
    def __init__(self, queue_name="mbt_testing_queue"):
        self.queue_name = queue_name
        self.host = "rabbitmq.core.svc.cluster.local"
        self.port = 5672
        self.user = os.getenv("AMQ_USER", "guest")
        self.password = os.getenv("AMQ_PASSWORD", "guest")
        self.connection = None
        self.channel = None

    def _connect(self):
        if self.connection and self.connection.is_open:
            return

        credentials = pika.PlainCredentials(self.user, self.password)
        parameters = pika.ConnectionParameters(self.host, self.port, "/", credentials)
        self.connection = pika.BlockingConnection(parameters)
        self.channel = self.connection.channel()

        self.channel.queue_declare(queue=self.queue_name, durable=True)

    def send_message_async(self, msg_type: str, proto_msg):
        def task():
            try:
                self._connect()
                payload = self._prepare_payload(msg_type, proto_msg)
                self.channel.basic_publish(
                    exchange="",
                    routing_key=self.queue_name,
                    body=payload,
                    properties=pika.BasicProperties(content_type="application/json")
                )
            except Exception as e:
                print(f"Error sending message to RabbitMQ: {e}")

        threading.Thread(target=task).start()

    def _prepare_payload(self, msg_type: str, proto_msg) -> str:
        binary_body = proto_msg.SerializeToString()
        encoded_body = base64.b64encode(binary_body).decode("utf-8")
        message = {
            "type": msg_type,
            "body": encoded_body
        }
        return json.dumps(message)
